be/src/service/service_be/lake_service.cpp

```diff
@@ -4,27 +4,146 @@
+        response->mutable_status()->add_error_msgs("missing txn_ids");
+        return;
+    }
+    if (request->tablet_ids_size() == 0) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("missing tablet_ids");
+        return;
+    }
+    int64_t base_version = request->version() - request->txn_ids_size();
+    if (base_version < 1) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("invalid version");
+        return;
+    }
+    // Will not update status code since here, only failed_tablets will be updated.
+    response->mutable_status()->set_status_code(TStatusCode::OK);
+
+    // TODO move the execution to TaskWorkerPool
+    for (const auto& tablet_id : request->tablet_ids()) {
+        auto tablet = _env->lake_tablet_manager()->get_tablet(tablet_id);
+        if (!tablet.ok()) {
+            LOG(WARNING) << "Fail to get tablet " << tablet_id << ": " << tablet.status();
+            response->add_failed_tablets(tablet_id);
+            continue;
+        }
+        auto base_metadata = tablet->get_metadata(base_version);
+        if (base_metadata.status().is_not_found() && tablet->get_metadata(request->version()).ok()) {
+            // the current tablet has been successfully published before, skip to the next tablet
+            continue;
+        } else if (!base_metadata.ok()) {
+            LOG(WARNING) << "Fail to get " << tablet->metadata_path(base_version) << ": " << base_metadata.status();
+            response->add_failed_tablets(tablet_id);
+            continue;
+        }
+
+        // make a copy of metadata
+        auto new_metadata = std::make_shared<lake::TabletMetadata>(**base_metadata);
+        new_metadata->set_version(request->version());
+
+        bool succeeded = false;
+        for (auto txn_id : request->txn_ids()) {
+            auto txnlog = tablet->get_txn_log(txn_id);
+            if (txnlog.status().is_not_found() && tablet->get_metadata(request->version()).ok()) {
+                // the current tablet has been successfully published before, skip to the next tablet
+                break;
+            } else if (!txnlog.ok()) {
+                LOG(WARNING) << "Fail to get " << tablet->txn_log_path(txn_id) << ": " << txnlog.status();
+                response->add_failed_tablets(tablet_id);
+                break;
+            }
+            // else
+            auto st = apply_txn_log(**txnlog, new_metadata.get());
+            if (!st.ok()) {
+                LOG(WARNING) << "Fail to apply " << tablet->txn_log_path(txn_id) << ": " << st;
+                response->add_failed_tablets(tablet_id);
+                break;
+            }
+            succeeded = true;
+        }
+        if (!succeeded) {
+            continue;
+        }
+        auto st = tablet->put_metadata(new_metadata);
```

Comment:
The case below may happen.
Only part of the txns apply successful and delete all txn logs at the end.

Reply:
Sorry, I didn't grasp your question.

Reply:
This won't happen in the current implementation.